<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import sqlalchemy as db
from sqlalchemy import text
from sqlalchemy import inspect 
import pandas as pd

In [2]:
# установка соединения


In [3]:
# Получение списка таблиц в базе данных и их столбцов с типами данных:

# Создание объекта inspector для анализа базы данных:
inspector = inspect(engine)

# Получение списка имен таблиц в базе данных:
table_names = inspector.get_table_names()

# Вывод информации о количестве таблиц в базе данных:
print(f"Кол-во таблиц в этой БД - {len(table_names)}.")

# Перебор и вывод информации о таблицах:   
for table_name in table_names:
    print(f"\t{table_name}:")
    print(f"Столбцы в таблице '{table_name}':")
    columns = inspector.get_columns(table_name)
    for column in columns:
        print(f"Название: {column['name']}\tТип: {column['type']}")
    print()

Кол-во таблиц в этой БД - 13.
	test:
Столбцы в таблице 'test':
Название: name	Тип: TEXT
Название: cnt	Тип: BIGINT

	t_tab1:
Столбцы в таблице 't_tab1':
Название: id	Тип: INTEGER
Название: goods_type	Тип: VARCHAR(255)
Название: quantity	Тип: INTEGER
Название: amount	Тип: INTEGER
Название: seller_name	Тип: VARCHAR(255)

	t_tab2:
Столбцы в таблице 't_tab2':
Название: id	Тип: INTEGER
Название: name	Тип: VARCHAR(255)
Название: salary	Тип: INTEGER
Название: age	Тип: INTEGER

	operations:
Столбцы в таблице 'operations':
Название: id	Тип: BIGINT
Название: operdate	Тип: DATE
Название: recieverid	Тип: BIGINT
Название: senderid	Тип: BIGINT
Название: amount	Тип: NUMERIC(15, 2)

	table_1:
Столбцы в таблице 'table_1':
Название: name	Тип: TEXT
Название: cnt	Тип: BIGINT

	films_recent:
Столбцы в таблице 'films_recent':
Название: name	Тип: TEXT
Название: count	Тип: BIGINT

	cities:
Столбцы в таблице 'cities':
Название: city	Тип: ARRAY

	employers:
Столбцы в таблице 'employers':
Название: id	Тип: INTEGE

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# SQL-запрос для получения количества вакансий в таблице "vacancies"
query3_1 = '''
    SELECT 
   	    COUNT (*) AS cnt_vacancies
    FROM 
        vacancies;
         '''

In [5]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

cnt_vacancies
0          49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# SQL-запрос для получения количества работодателей в таблице "employers"
query3_2 = '''
     SELECT 
   	     COUNT (employers.id) AS cnt_employers
     FROM 
          employers;
         '''

In [7]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_2)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

cnt_employers
0          23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# SQL-запрос для получения количества регионов в таблице "areas"
query3_3 = '''
     SELECT 
   	     COUNT (areas.id) AS cnt_areas
     FROM 
          areas;
         '''

In [9]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_3)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

cnt_areas
0       1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# SQL-запрос для получения количества сфер деятельности в таблице "industries"
query3_4 = '''
     SELECT 
   	     COUNT (id) AS cnt_industries
     FROM 
          industries;
         '''

In [11]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_4)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

cnt_industries
0             294

Дополнительное исследование 1: Среднее количество вакансий на одного работодателя

In [12]:
# SQL-запрос для определния среднего количества вакансий на одного работодателя в таблице vacancies
query3_A1 = '''
     SELECT 
   	     (COUNT (DISTINCT id)/COUNT(DISTINCT employer_id)) AS avg_vacancies
     FROM 
          vacancies;
         '''

In [13]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_A1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

avg_vacancies
0              3

Дополнительное исследование 2: Среднее количество сфер деятельности на одного работодателя

In [14]:
# SQL-запрос для определения среднего количества сфер деятельности на одного работодателя в таблице EMPLOYERS_INDUSTRIES
query3_A2 = '''
     SELECT 
   	     (COUNT (industry_id)/COUNT(DISTINCT(employer_id))) AS avg_industries
     
     FROM 
          EMPLOYERS_INDUSTRIES;
         '''

In [15]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query3_A2)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

avg_industries
0               2

# выводы по предварительному анализу данных

Общее количество вакансий в таблице vacancies - 49197. Общее количество работодателей employers - 23501. Количество регионов в таблице areas - 1362. Общее количество сфер деятельности в таблице industries - 294.

Из полученных данных следует, что в среднем, на одного работодателя приходится по 3 вакансии. В среднем, на одну сферу деятельности работодателя приходятся  в 2-х разных сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [16]:
# SQL запрос для определения количества вакансий в каждом регионе с сортировкой по кол-ву вакансий в порядке убывания
query4_1 = '''
    SELECT 
        a.name AS area,
        count(v.id) AS cnt
    FROM 
        vacancies AS v
    JOIN 
        areas AS a ON a.id = v.area_id
    GROUP BY 
        a.name
    ORDER BY 
        2 DESC;    
         '''


In [17]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [18]:
# SQL запрос, определяющий количество вакансий где заполнено хотя бы одно из двух полей с зарплатой
query4_2 = '''
SELECT 
    COUNT(v.id)
FROM 
    vacancies AS v
WHERE 
    v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL;
    '''

In [19]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_2)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [20]:
# текст запроса
query4_3 = '''
    SELECT
        'AVG minimum salary' :: text AS salary,
        ROUND(AVG(salary_from)) AS amount
    FROM vacancies
    UNION
    SELECT
        'AVG maximum salary' :: text AS salary,
        ROUND(AVG(salary_to)) AS Amount
    FROM vacancies
    UNION
    SELECT
        'Salary spread' :: text AS salary,
        ROUND(AVG(salary_to)) - ROUND(AVG(salary_from)) AS Amount
    FROM vacancies
    UNION
    SELECT
        '% spread to min salary' :: text AS salary,
        ROUND((ROUND(AVG(salary_to)) - ROUND(AVG(salary_from))) / ROUND(AVG(salary_from)) * 100, 2) AS Amount
    FROM vacancies
    ORDER BY amount;                 
        '''

In [21]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_3)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

salary  amount
0  % spread to min salary   55.54
1           Salary spread   39472
2      AVG minimum salary   71065
3      AVG maximum salary  110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [22]:
# SQL запрос, выводящий кол-во вакансий 
# для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment). 
# Результат отсортирован по убыванию количества:
query4_4 = '''
    SELECT
        schedule,
        employment,
        COUNT(id) AS nb_vacancies
    FROM
        vacancies
    GROUP BY 
        schedule, employment
    ORDER BY 
        nb_vacancies DESC;
'''

In [23]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_4)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

schedule           employment  nb_vacancies
0        Полный день     Полная занятость         35367
1   Удаленная работа     Полная занятость          7802
2      Гибкий график     Полная занятость          1593
3   Удаленная работа  Частичная занятость          1312
4     Сменный график     Полная занятость           940
5        Полный день           Стажировка           569
6     Вахтовый метод     Полная занятость           367
7        Полный день  Частичная занятость           347
8      Гибкий график  Частичная занятость           312
9        Полный день     Проектная работа           141
10  Удаленная работа     Проектная работа           133
11     Гибкий график           Стажировка           116
12    Сменный график  Частичная занятость           101
13  Удаленная работа           Стажировка            64
14     Гибкий график     Проектная работа            18
15    Сменный график           Стажировка            12
16    Вахтовый метод     Проектная работа             2
17    Сменный график     Проектная работа             1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [24]:
# SQL запрос, выводящий значения поля Требуемый опыт работы (experience)
# в порядке возрастания количества вакансий, в которых указан данный вариант опыта:
query4_5 = '''
    SELECT
        experience,
        COUNT(id) AS nb_vacancies
    FROM 
        vacancies
    GROUP BY 
        experience
    ORDER BY 
        nb_vacancies ASC;
'''

In [25]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_5)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

experience  nb_vacancies
0         Более 6 лет          1337
1           Нет опыта          7197
2       От 3 до 6 лет         14511
3  От 1 года до 3 лет         26152

Дополнительное исследование 1: Регионы по разнообразию сфер деятельности

In [26]:
# SQL-запрос для ранжирования регионов по разнообразию сфер деятельности
query4_A1 = '''
     SELECT 
          a.name AS area,
          COUNT (DISTINCT v.employer_id) AS nb_employers,
          COUNT (DISTINCT ei.industry_id) AS nb_industries
     FROM 
          areas AS a
     LEFT JOIN
          vacancies AS v ON a.id = v.area_id
     LEFT JOIN
          employers_industries AS ei ON v.employer_id = ei.employer_id
     GROUP BY
          area
     ORDER BY
          nb_industries DESC;
         '''

In [27]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query4_A1)).fetchmany(10)
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

area  nb_employers  nb_industries
0           Москва          2320            270
1  Санкт-Петербург          1451            255
2  Нижний Новгород           657            214
3     Екатеринбург           742            214
4      Новосибирск           802            212
5   Ростов-на-Дону           487            199
6        Краснодар           578            197
7           Алматы           798            196
8           Самара           473            191
9           Казань           609            180

***

## Выводы по детальному анализу вакансий ##
### Регионы по количеству вакансий ###

Топ 10 регионов по количеству вакансий - Москва	(5333), Санкт-Петербург	(2851),	Минск (2112), Новосибирск (2006), Алматы (1892), Екатеринбург	(1698), Нижний Новгород	(1670),	Казань (1415), Краснодар (1301), Самара (1144). Города, входящие в этот список (разумеется, кроме столиц Республики Беларусь и Республики Казахстан), представляют собой крупнейшие административные, промышленные и финансовые центры как своих регионов, так и РФ в общем. Кроме того, необходимо отметить значительный отрыв (почти в 2,5 раза) Москвы по количеству вакансий от следующего города - Санкт-Петербурга при том, что разрыв между всеми последующими городами является значительно меньшим.

### Зарплатные предложения ###
Количество вакансий с указанием зарплаты (24073) составляет менее половины от общего числа опубликованых вакансий (49197).
Среднее значение максимальной зарплаты составляет 110537 руб. Среднее значение минимальной зарплаты - 71065 руб. Спрэд зарплатных предложений составляет 39472 руб. Средняя максимальная зарплата на 55,54 % выше средней минимальной зарплаты. 

### Распределение вакансий по типу занятости и графика ###
Количество вакансий для каждого типа занятости - наибольшее количество вакансий приходится на полную занятость. 
Первое место - у сочетания полная занятость и полный рабочий день, затем полная занятость - гибкий график,
на третьем месте - гибкий график. Наименьшее количество вакансий по типу занятости проектная работа, 
Bottom-3 таблицы представляет собой, на последнем месте, проектная работа - сменный график,
над ним - проектная работа - вахтовый метод, на третьем снизу месте - стажировка - сменный график. Это говорит о том, что организация труда на большинстве предприятий не предполагает гибкости работы, что может быть связанно либо с особенностями производственного процесса на предприятии (особенно касается реального сектора экономики и ряда секторов сферы услуг), либо с привычкой как работодателей, так и соискателей, мало изменившейся несмотря на опыт периода ковидных ограничений.  

### Требования к опыту работы ###
Больше всего вакансий относительно требований к опыту работы приходится на опыт от 1 до 3-х лет (26152), затем, почти вдвое меньше, 
вакансий с требованием к опыту работы 3-6 лет (14511), меньше всего вакансий для людей с опытом работы свыше 6 лет. О причинах такого распределения сложно судить, но, скорее всего, предприятия предпочитают набирать на работу молодых квалифицированных сотрудников с изначально меньшими зарплатными требованиями, которых можно легче обучить работе в условиях конкретного предприятия и легче встроить в уже существующие производственные процессы.

### Регионы по разнообразию сфер деятельности работодателей ###
Прежде всего, необходимо уточнить, что разнообразие сфер деятельности описывает работодателей а не вакансии, так как в самих вакансиях сфера деятельности не указывается. Однако, косвенно это также помогает определить и разнообразие сфер деятельности в предлагаемых вакансиях.
Позиции регионов по разнообразию сфер деятельности работодателей, публикующих объявления, несколько отличается от топ-10 по количеству вакансий. Первые 2 места по-прежнему занимают Москва (270 из 294-х существующих сфер деятельности) и Санкт-Петербург (255), за которыми следуют Нижний Новгород и Екатеринбург (214), затем Новосибирск (212), Ростов-на-Дону (199), Краснодар (197), столица Республики Казахстан Алматы (196), Самара (191), Казань (180). При этом необходимо отметить, что это разнообразие сфер деятельности не коррелирует с количеством работодателей в регионе.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [28]:
# SQL запрос, позволяющий узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.
query5_1 = '''
    SELECT
        e.name AS employer_name,
        COUNT(v.id) AS nb_vacancies 
    FROM 
        employers AS e
    JOIN vacancies AS v ON e.id = v.employer_id
    GROUP BY 
        e.name
    ORDER BY 
        nb_vacancies DESC
    LIMIT 5;
'''

In [29]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

employer_name  nb_vacancies
0         Яндекс          1933
1     Ростелеком           491
2       Тинькофф           444
3           СБЕР           428
4  Газпром нефть           331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [30]:
# SQL запрос, который для каждого региона выведет количество работодателей и вакансий в нём. 
# Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.
query5_2 = '''
    SELECT
        a.name AS area,
        COUNT(DISTINCT e.id) AS nb_employers,
        COUNT(v.id) AS nb_vacancies
    FROM 
        areas AS a
    LEFT JOIN employers AS e ON e.area = a.id
    LEFT JOIN vacancies AS v ON v.area_id = a.id
    WHERE 
        v.id IS NULL
    GROUP BY 
        a.name
    ORDER BY 
        nb_employers DESC;
'''

In [31]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_2)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

area  nb_employers  nb_vacancies
0                            Россия           410             0
1                         Казахстан           207             0
2                Московская область            75             0
3                Краснодарский край            19             0
4                          Беларусь            18             0
..                              ...           ...           ...
588                        Кувандык             0             0
589                        Куйбышев             0             0
590  Куйбышево (Ростовская область)             0             0
591                        Култаево             0             0
592                        Кульсары             0             0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [32]:
# SQL запрос, считающий количество регионов где каждый работодатель публикует свои вакансии
# и сортирующий результат по убыванию количества:
query5_3 = '''
    SELECT 
        e.name AS employer,
        COUNT(DISTINCT area_id) AS nb_area
    FROM
        vacancies AS v 
    JOIN employers AS e ON v.employer_id = e.id
    GROUP BY 
        e.name 
    ORDER BY 
        nb_area DESC;
'''

In [33]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_3)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

employer  nb_area
0                      Яндекс      181
1                  Ростелеком      152
2                  Спецремонт      116
3      Поляков Денис Иванович       88
4                   ООО ЕФИН        71
...                       ...      ...
14761                  UniSol        1
14762            UNISTORY LLC        1
14763                   UNIT6        1
14764     United Distribution        1
14765              UNITEL LLC        1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [34]:
# SQL запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.
query5_4 = '''
    SELECT
        COUNT(DISTINCT(e.id)) AS nb_employers
    FROM
        employers AS e
    LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id 
    WHERE 
        ei.industry_id IS NULL;
'''

In [35]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_4)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

nb_employers
0          8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [36]:
# SQL запрос, выводящий название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, 
# у которых указано четыре сферы деятельности.
query5_5 = '''
    SELECT
        e.name
    FROM
        employers AS e
    LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id 
    GROUP BY 
        e.name
    HAVING 
        COUNT(ei.industry_id) = 4
    ORDER BY 
        e.name
    LIMIT 3;
'''

In [37]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_5)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [38]:
# SQL-запрос для определения кол-ва работодателей со сферой деятельности "Разработка программного обеспечения"
query5_6 = '''
    SELECT
        COUNT(DISTINCT(ei.employer_id))
    FROM
        employers_industries as ei
    LEFT JOIN industries AS i ON ei.industry_id = i.id
    WHERE 
        i.name = 'Разработка программного обеспечения';
'''

In [39]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_6)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [40]:
# Приведение URL к виду, удобному для запроса страницы из Wikipedia с помощью библиотеки urllib:
import urllib.parse as req 

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

In [41]:
# Помещение в кортеж названий городов, находящихся в таблице No. 1 (города-миллионеры)
# и перечисленных в столбце с индексом 1 (названия городов):
mil_cities = tuple(pd.read_html(url)[1].iloc[:, 1])
print(mil_cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [42]:
# SQL-запрос выводящий список регионов-миллионников для компании "Яндекс" 

query5_7 = f'''
    WITH yandex AS (                                        -- Создание подзапроса для упрощения общей структуры запроса
        SELECT 
            a.name AS city_name,
            COUNT(v.id) AS nb_vacancies
        FROM 
            areas AS a
        LEFT JOIN vacancies AS v ON a.id = v.area_id
        LEFT JOIN employers AS e ON v.employer_id = e.id
        WHERE 
            a.name IN {mil_cities} AND e.name = 'Яндекс'    -- Подключение к поиску кортежа полученного из Википедии
        GROUP BY 
            city_name
        ORDER BY 
            nb_vacancies)

    SELECT 
        *
    FROM 
        yandex                                              -- Обращение к подзапросу описанному выше
    UNION ALL
    SELECT
        'Total', SUM (nb_vacancies)                         -- Добавление строки "Сумма" 
    FROM 
        yandex;
'''

In [43]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query5_7)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

city_name nb_vacancies
0              Омск           21
1         Челябинск           22
2        Красноярск           23
3         Волгоград           24
4             Пермь           25
5            Казань           25
6    Ростов-на-Дону           25
7               Уфа           26
8            Самара           26
9         Краснодар           30
10          Воронеж           32
11      Новосибирск           35
12  Нижний Новгород           36
13     Екатеринбург           39
14  Санкт-Петербург           42
15           Москва           54
16            Total          485

***

# выводы по анализу работодателей
Топ-5 работодателей по количеству вакансий среди всех работодателей - Яндекс (1933), Ростелеком (491), Тинькофф	(444), СБЕР	(428), Газпромнефть (331). Внимание на себя обращает тот факт, что четыре из 5 работодателей являются компаниями с очень развитой IT составляющей.
По каоличеству работодателей среди регионов, в которых при отсутствуют вакансии при наличии работодателей, на первом месте РФ, затем Республика Казахстан, Московская область, Краснодарский край и Республика Беларусь. Необходимо уточнить, что взаимное сравнение образований государственного, регионального, областного, городского и муниципального масштаба не может быть информативным.
Компании, которые присутствуют в наибольшем количестве регионов - Яндекс (181), Ростелеком (152) и Спецремонт (116) - также имеют явно выраженную IT направленность.

У значительной части работодателей - 8419 из 23501 - не указана сфера деятельности.
Топ 3 работодателей, для которых указаны 4 сферы деятельности состоит из IT компаний.
Из 23501 работодателя 3553 указали "Разработку програмного обеспечения" в качестве одной из своих сфер деятельности.

Компания "Яндекс", которая является лидером среди работодателей по количеству вакансий и их распределённости, предлагает 485 из 1933-х вакансий в городах-миллионерах РФ. Первые два места занимают Москва (54) и Санкт-Петербург (42), занимающие первые 2 места в любых количественных рейтингах рассмотренных ранее. Топ-3 по количеству вакансий "Яндекс" совпадает с Топ-3 по разнообразию сфер деятельности работодателей в регионе.





# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [44]:
# SQL-запрос для определения вакансий, относящихся к данным:
query6_1 = '''
    SELECT
        COUNT(id)
    FROM 
        vacancies
    WHERE 
        LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%';     -- Приведение данных в нижний регистр для повышения точности поиска
'''

In [45]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [46]:
# SQL-запрос поиска вакансий по описанным выше условиям:
query6_2 = '''
SELECT
    COUNT(id)
FROM vacancies
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%')
            
    AND (LOWER(name) LIKE '%junior%' 
        OR LOWER(experience) LIKE 'Нет опыта' 
            OR LOWER(employment) LIKE 'Стажировка');      
'''

In [47]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_2)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0     35

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [48]:
# SQL-запрос для поиска вакансий с указанием SQL или postgres:  
query6_3 = '''
SELECT
    COUNT(v.id)
FROM vacancies AS v
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%')
                        
    AND (v.key_skills LIKE '%SQL%' OR v.key_skills LIKE '%postgres%');            
'''

In [49]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_3)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [50]:
# SQL-запрос для проверки популярности Python в требованиях к соискателям:
query6_4 = '''
SELECT
    COUNT(DISTINCT v.id)
FROM vacancies AS v
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%')
            
    AND (key_skills LIKE '%Python%');            
'''

In [51]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_4)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [52]:
# SQL-запрос для подсчёта среднего числа ключевых навыков в требованиях к DS соискателям:
query6_5 = '''
SELECT
    ROUND(AVG((LENGTH(key_skills) - LENGTH(REPLACE(key_skills,'\t',''))+1)),2)      --Рассчёт кол-ва навыков по принципу кол-во разделителей +1
FROM vacancies AS v
WHERE 
     (LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%');
'''

In [53]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_5)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

round
0  6.41

Дополнительное исследование 6.5.1: Каково среднее количество требуемых навыков среди всех предложений

In [61]:
# SQL-запрос для подсчёта среднего числа ключевых навыков в требованиях к соискателям:
query6_5_1 = '''
SELECT
    ROUND(AVG((LENGTH(key_skills) - LENGTH(REPLACE(key_skills,'\t',''))+1)),2)      --Рассчёт кол-ва навыков по принципу кол-во разделителей +1
FROM vacancies AS v
WHERE NOT
     ((LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%'));
'''

In [62]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_5_1)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

round
0  6.37

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# SQL-запрос для рассчёт средней зарплаты в зависимости от требуемого опыта:
query6_6 = '''
SELECT
    experience,
    ROUND(AVG((COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2)) AS mean_salary
FROM 
    vacancies
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
            OR LOWER(name) LIKE '%machine learning%' OR LOWER(name) LIKE '%машинн%обучен%')
            
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY 
    experience;
'''

In [ ]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_6)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

experience mean_salary
0           Нет опыта       74643
1  От 1 года до 3 лет      139675
2       От 3 до 6 лет      243115

In [ ]:
# Выполнение SQL-запроса и получение результата:
result = conn.execute(text(query6_6)).fetchall()
# Создание DataFrame из результата:
df = pd.DataFrame(result)
df

# выводы по предметному анализу

В ходе исследования была обнаружена 1771 вакансия, имеющая отношение к данным и работе с ними. Из них подходящих для начинающего data scientist - 35.
Всего среди вакансий для DS требуют владения SQL или postgres - 201
Владение Python предполагает 351 вакансия
Среднее количество необходимых навыков, указанных в DS вакансиях - 6.41

Средняя зарплата в сфере DS значительно возрастает с увеличением опыта работы соискателя. Без опыта работы прелагаемая зарплата 
в среднем равна 74643, при опыте работы 1-3 года она составляет в среднем 139675, затем, при опыте работы 3-6 лет - 243115. Эти уровни зарплатных предложений явно выше средних по рынку (71065 - 110537 руб).

# Общий вывод по проекту

## Основные выводы полученные на основе промежуточных результатов ##

Рынок труда представлен большим количеством вакансий и разнообразием сфер деятельности и регионов.
Среднее количество вакансий на одного работодателя говорит о наличии конкуренции между работодателями.

Москва и крупные города лидируют по количеству вакансий, при этом Москва имеет значительный отрыв от других городов.

Предприятия, публикующие вакансии, часто не предоставляют всю необходимую информацию, что затрудняет выбор вакансий.
Меньше половины вакансий указывают информацию о зарплате, что явно является неудобным для соискателя. Значительное количество работодателей не указали сферу деятельности компаний, что также может вызывать сложности для тех соискателей, которые хотят запланировать свой карьерный путь в рамках одной компании. Подобная типичная структура объявлений о вакансии говорит о том, что речь идёт о рынке труда "сильного работодателя", где преимущество переговорной позиции изначально находится на стороне работодателя.

Опыт работы от 1 до 3 лет наиболее распространен среди требований к соискателям, что может говорить о том, что компании предпочитают набирать молодых специалистов и доучивать их согласно специфике каждого отдельного предприятия. Кроме того, можно предположить, что текучесть более опытных кадров значительно ниже и, соответственно, вакансии для них освобождаются реже. Это так же может говорить о том, что мобильность более опытных кадров обеспечивается в основном другими средствами, помимо публикации вакансии на сайте HH. Кроме того, малый спрос на опытные кадры является признаком того, что предприятия не расширяют объёмы своей работы, не открывают новых торговых или промышленных мощностей и, соответственно, не нуждаются в увеличении количества более опытных и ответственных сотрудников для низовой организации такого рода расширения. 

IT-компании, такие как Яндекс и Ростелеком, имеют высокое количество вакансий и разнообразие сфер деятельности в разных регионах. Наибольшее количество вакансий, со значительным отрывом, приходится на Москву, что говорит о значительной централизации и географической разбалансированности распределения занятости в IT.

Существует достаточно ограниченное количество вакансий для Data Scientist, и большинство из них требуют опыта и специфических навыков, таких как SQL и Python. Среднее количество требуемых навыков для вакансий в Data Scientist незначительно превышает аналогичное среднее требование по рынку. 

Зарплата Data Scientist сильно зависит от опыта работы но в большинстве случаев значительно превышает среднюю максимальную зарплату по рынку.


### Варианты продолжения исследования: ###

Анализ зарплатных трендов: Проведение более подробного анализа зарплатных данных для разных регионов и сфер деятельности, чтобы выявить тенденции и сравнить их с текущими стандартами рынка.

Исследование влияния COVID-19: Анализ изменений в рынке труда, вызванных пандемией COVID-19, и прогнозирование долгосрочных последствий для различных отраслей и регионов.

Моделирование спроса и предложения: Разработка моделей для прогнозирования спроса на рабочую силу и предложения вакансий на рынке труда.

Анализ качества соискателей: Исследование качества соискателей, их образования и навыков, чтобы лучше соотнести предложение и спрос на рынке труда.
 
Мониторинг рынка труда: Регулярное обновление данных и мониторинг изменений на рынке труда для адаптации к динамике спроса и предложения.

Специализированный анализ отраслей: Проведение более детального анализа конкретных отраслей и регионов для выявления уникальных тенденций и потребностей.

Социальный анализ: Исследование вопросов включения, равенства и разнообразия на рынке труда для создания более справедливой рабочей среды.

Экономико-географический анализ: Рынок труда как индикатор изменений в географическом распределении ключевых отраслей экономики и экономического здоровья регионов.